### ISSUE TITLE
Conversion broken with rpy2 3.4.3

### ISSUE THREAD

**Issue by dburkhardt** [theislab/anndata2ri#63]
After updating rpy2 to version 3.4.3 anndata2ri breaks

```python
import scanpy as sc
import anndata2ri

adata = sc.datasets.pbmc3k()
anndata2ri.py2rpy(adata)
```

raises

```pytb
---------------------------------------------------------------------------
RRuntimeError                             Traceback (most recent call last)
<ipython-input-12-d1c9812891ee> in <module>
----> 1 anndata2ri.py2rpy(adata)

/opt/miniconda3/lib/python3.8/site-packages/anndata2ri/__init__.py in py2rpy(obj)
     46         - :class:`~anndata.AnnData` → :rcls:`~SingleCellExperiment::SingleCellExperiment`
     47         """
---> 48         return converter.py2rpy(obj)
     49
     50     def rpy2py(obj: Any) -> Sexp:

/opt/miniconda3/lib/python3.8/functools.py in wrapper(*args, **kw)
    873                             '1 positional argument')
    874
--> 875         return dispatch(args[0].__class__)(*args, **kw)
    876
    877     funcname = getattr(func, '__name__', 'singledispatch function')

/opt/miniconda3/lib/python3.8/site-packages/anndata2ri/py2r.py in py2rpy_anndata(obj)
     54         sce = importr("SingleCellExperiment")
     55         # TODO: sparse
---> 56         x = {} if obj.X is None else dict(X=mat_converter.py2rpy(obj.X.T))
     57         layers = {k: mat_converter.py2rpy(v.T) for k, v in obj.layers.items()}
     58         assays = ListVector({**x, **layers})

/opt/miniconda3/lib/python3.8/functools.py in wrapper(*args, **kw)
    873                             '1 positional argument')
    874
--> 875         return dispatch(args[0].__class__)(*args, **kw)
    876
    877     funcname = getattr(func, '__name__', 'singledispatch function')

/opt/miniconda3/lib/python3.8/site-packages/anndata2ri/scipy2ri/py2r.py in wrapper(obj)
     41
     42         with localconverter(default_converter + numpy2ri.converter):
---> 43             return f(obj)
     44
     45     return wrapper

/opt/miniconda3/lib/python3.8/site-packages/anndata2ri/scipy2ri/py2r.py in csc_to_rmat(csc)
     51     csc.sort_indices()
     52     t, conv_data, _ = get_type_conv(csc.dtype)
---> 53     return methods.new(
     54         f"{t}gCMatrix",
     55         i=as_integer(csc.indices),

/opt/miniconda3/lib/python3.8/site-packages/rpy2/robjects/functions.py in __call__(self, *args, **kwargs)
    196                 v = kwargs.pop(k)
    197                 kwargs[r_k] = v
--> 198         return (super(SignatureTranslatedFunction, self)
    199                 .__call__(*args, **kwargs))
    200

/opt/miniconda3/lib/python3.8/site-packages/rpy2/robjects/functions.py in __call__(self, *args, **kwargs)
    123             else:
    124                 new_kwargs[k] = conversion.py2rpy(v)
--> 125         res = super(Function, self).__call__(*new_args, **new_kwargs)
    126         res = conversion.rpy2py(res)
    127         return res

/opt/miniconda3/lib/python3.8/site-packages/rpy2/rinterface_lib/conversion.py in _(*args, **kwargs)
     43 def _cdata_res_to_rinterface(function):
     44     def _(*args, **kwargs):
---> 45         cdata = function(*args, **kwargs)
     46         # TODO: test cdata is of the expected CType
     47         return _cdata_to_rinterface(cdata)

/opt/miniconda3/lib/python3.8/site-packages/rpy2/rinterface.py in __call__(self, *args, **kwargs)
    678             )
    679             if error_occured[0]:
--> 680                 raise embedded.RRuntimeError(_rinterface._geterrmessage())
    681         return res
    682

RRuntimeError: Error in validObject(.Object) :
  invalid class “dgCMatrix” object: 1: invalid object for slot "i" in class "dgCMatrix": got class "array", should be or extend class "integer"
invalid class “dgCMatrix” object: 2: invalid object for slot "p" in class "dgCMatrix": got class "array", should be or extend class "integer"
invalid class “dgCMatrix” object: 3: invalid object for slot "Dim" in class "dgCMatrix": got class "array", should be or extend class "integer"
invalid class “dgCMatrix” object: 4: invalid object for slot "x" in class "dgCMatrix": got class "array", should be or extend class "numeric"
```

**Comment by flying-sheep** [comment]
Hmm, probably rpy2/rpy2#786.

**Comment by LuckyMD** [comment]
Hey @flying-sheep! I don't think this is related to rpy2/rpy2#786 as that was fixed for 3.4.3, which is the very in which this error occurs.

**Comment by flying-sheep** [comment]
Exactly. They might call it a fix, while it breaks things here.

**Comment by LuckyMD** [comment]
Ah, I see. Do you know where this issue occurs on the `rpy2` side? Or rather in what `rpy2` function that `anndata2ri` relies on? Would be useful to report the issue there.

**Comment by eijynagai** [comment]
Hi, 

Same here, I tried this code running in Jupyter notebook (python3 kernel): 
```python
%%R -o so_sce
so_sce <- as.SingleCellExperiment(so)
```
 After changing to the last update (rpy2 3.4.4) I started getting exactly the same error as @dburkhardt.

**Comment by eijynagai** [comment]
Hi again, 

JFYI, I downgraded to version 3.4.2 and it worked. I am not sure where the issue is...
```python
pip install --force-reinstall 'rpy2==3.4.2'
```

**Comment by liuzj039** [comment]
Same issue.
But when I downgraded my rpy2 to 3.3.6 version. This problem were solved.

**Comment by galicae** [comment]
Same issue for me - I tried downgrading rpy2 to 3.4.2, 3.3.6 and others, to no avail. The minimal example breaks with the same exact error:

```
import scanpy as sc
import anndata2ri

adata = sc.datasets.pbmc3k()
anndata2ri.py2rpy(adata)
```

The relevant package versions:

```
anndata     0.7.6
scanpy      1.7.2
anndata2ri          1.0.6
rpy2                3.4.5
```

**Comment by liuzj039** [comment]
@galicae 
If you are a conda user, may be you can uninstall rpy2 by conda first, and then install the 3.4.2 version by pip.
This solution has been solve a lot of my issues about incompatibility.

**Comment by galicae** [comment]
OMG that did the trick, thank you so much @liuzj039! Really frustrating couple of days trying to figure this one out...

**Comment by fabianrost84** [comment]
A workaround is to make the expression matrix dense. This works:
```python
import scanpy as sc
import anndata2ri

adata = sc.datasets.pbmc3k()
adata.X = adata.X.A
anndata2ri.py2rpy(adata)
```

**Comment by liuzj039** [comment]
Another option is to simply use the following code in place of this function. https://github.com/theislab/anndata2ri/blob/8a1fdd431eeca5443a50f79c6de293210fb30699/anndata2ri/scipy2ri/py2r.py#L31

```python
def _py2r(x):
    import rpy2.robjects as ro
    with localconverter(
        default_converter
        + numpy2ri.converter
    ):
        x = ro.conversion.py2rpy(x)
    return x


def py2r_context(f):
    @wraps(f)
    def wrapper(obj):
        global methods, as_logical, as_integer, as_double
        if methods is None:
            importr("Matrix")  # make class available
            methods = importr("methods")
            as_logical = lambda x:baseenv["as.logical"](_py2r(x))
            as_integer = lambda x:baseenv["as.integer"](_py2r(x))
            as_double = lambda x:baseenv["as.double"](_py2r(x))

        return f(obj)

    return wrapper
```

**Comment by LuckyMD** [comment]
@flying-sheep is this a fix you could add into `anndata2ri`? Or is the Matrix R dependency a problem?

**Comment by LuckyMD** [comment]
@liuzj039 @fabianrost84 Does this still happen with rpy2 3.4.5? I recently got the conversion to work seamlessly with that.

**Comment by liuzj039** [comment]
@LuckyMD 
I created a new environment and used the latest versions of rpy2 (3.4.5) and anndata2ri (1.0.6).
Unfortunately, the following script triggers the same error if I don't change any of the anndata2ri code.
```python
import scipy.sparse as ss
import anndata2ri
x = ss.csr_matrix((3,3))
x = anndata2ri.scipy2ri.py2rpy(x)
```
Is it possible that the reason you didn't trigger this error is that you anndata doesn't contain a sparse matrix?

**Comment by LuckyMD** [comment]
Hmm... I did convert a sparse matrix from Seurat via SCE. But I only noticed that everything worked about a week afterwards... so I didn't think of this error and now don't recall which dataset that was.

**Comment by liuzj039** [comment]
Yes, you can transfer a sparse matrix from r to python, but the reverse does not work.

**Comment by LuckyMD** [comment]
ah, sorry about that ^^.

**Comment by liuzj039** [comment]
The cause of this problem seems to be that rpy2 always converts intvector to numpy.array after 3.4.3.  So some times the index of a sparse matrix that has already been converted is converted back to python again. 

Many thanks for quick reply !

**Comment by gokceneraslan** [comment]
Damn, downgrading rpy2 to 3.4.2 requires entire the bioconda stack to be downgraded, which is horrible. Does anyone have time to make a reproducible example and file a report on rpy2 github maybe 😇 ?

**Comment by flying-sheep** [comment]
@liuzj039 could you please do an anndata2ri PR?

**Comment by flying-sheep** [comment]
Hi! We finally have tests thanks to @scottgigante-immunai! Unfortunately they revealed what I suspected: That it wouldn’t be so easy. See #66

**Comment by flying-sheep** [comment]
Fixed by #80

**Comment by ngvananh2508** [comment]
Hi! Can anyone please help me with this issue?
When I import rpy2, it will success but when I import rpy2.robjects, it will fall and it appeared this error: RRuntimeError: Error in gettext(fmt, domain = domain, trim = trim) : 
  3 arguments passed to .Internal(gettext) which requires 2.
  And this error also happened with anndata2ri importing.
  
  Now I am running R 4.2.0, and python 3.10.9.
